File: src/main/java/zielu/gittoolbox/cache/PerRepoInfoCache.java
package zielu.gittoolbox.cache;

import com.intellij.openapi.project.Project;
import com.intellij.util.messages.Topic;
import git4idea.repo.GitRepository;
import org.jetbrains.annotations.NotNull;
import zielu.gittoolbox.util.AppUtil;

public interface PerRepoInfoCache extends DirMappingAware, RepoChangeAware {
  Topic<PerRepoStatusCacheListener> CACHE_CHANGE = Topic.create("Status cache change",
      PerRepoStatusCacheListener.class);
  @NotNull
  RepoInfo getInfo(GitRepository repository);

  void refreshAll();

  void refresh(Iterable<GitRepository> repositories);

  @NotNull
  static PerRepoInfoCache getInstance(@NotNull Project project) {
    return AppUtil.getServiceInstance(project, PerRepoInfoCache.class);
  }
}

File: src/main/kotlin/zielu/gittoolbox/cache/PerRepoInfoCache.kt
package zielu.gittoolbox.cache

import com.intellij.openapi.project.Project
import com.intellij.util.messages.Topic
import git4idea.repo.GitRepository
import zielu.gittoolbox.util.AppUtil.getServiceInstance

internal interface PerRepoInfoCache : DirMappingAware, RepoChangeAware {
  fun getInfo(repository: GitRepository): RepoInfo

  fun getAllInfos(): List<RepoInfo>

  fun refreshAll()

  fun refresh(repositories: Iterable<GitRepository>)

  companion object {
    @JvmField
    val CACHE_CHANGE_TOPIC = Topic.create("Status cache change", PerRepoStatusCacheListener::class.java)

    @JvmStatic
    fun getInstance(project: Project): PerRepoInfoCache {
      return getServiceInstance(project, PerRepoInfoCache::class.java)
    }
  }
}
